cerebrum/linalg/include/tensor.h

```diff
@@ -0,0 +1,102 @@
+#ifndef TENSOR_H_
+#define TENSOR_H_
+
+#include <array>
+#include <Eigen/Dense>
+#include <linalg/dim.h>
+#include <sstream>
+#include <stdexcept>
+#include <vector>
+
+namespace linalg {
+  template<typename T>
+  class Tensor {
+    public:
+      T* data;
+      dim_t dimension;
+
+      Tensor(int x, int y, int z) {
+        data = std::array<T, (x * y * z)>
+        dim.x = x;
+        dim.y = y;
+        dim.z = z;
+      }
+
+      Tensor(const Tensor& t) {
+        data = std::array<T, (t.x * t.y * t.z)>;
+        data = t.data;
+        dimension = t.dimension;
+      }
+
+      Tensor<T> operator+(const Tensor& t) {
+        Tensor<T> clone(*this);
+        for (int i = 0; i < dim_product(t); ++i) {
+          clone.data[i] += t.data[i];
+        }
+
+        return clone;
+      }
+
+      Tensor<T> operator-(const Tensor& t) {
+        Tensor<T> clone(*this);
+        for (int i = 0; i < dim_product(t); ++i) {
+          clone.data[i] -= t.data[i];
+        }
+
+        return clone;
+      }
+
+      T& get(int x, int y, int z) {
+        if (x < 0 || y < 0 || z < 0) {
+          throw new std::argument_exception("Error input dimensions cannot be < 0");
+        }
+        if (x > dimension.x || y > dimension.y || z > dimension.z) {
+          throw new std::argument_exception("Error, input dimensions cannot be larger than existing dimensions");
+        }
+
+        for (int i = 0; i < x; ++i) {
```

Comment:
This is kinda scary! Haha. Is O(n^3) necessary? Could we use a hash table or some other approach to reduce run time? I could be wrong. This may be the best way to avoid high space complexity or it just may be flat out necessary. Just looks unsettling at first sight. Feel free to ignore if this doesn't apply.

Reply:
Unfortunately yes, we need a way to copy the data into it. ML is quite intensive...